sale_triple_discount/tests/test_sale_triple_discount.py

```diff
@@ -0,0 +1,158 @@
+        self.so_line2.unlink()
+        # Divide by two on every discount:
+        self.so_line1.discount = 50.0
+        self.so_line1.discount2 = 50.0
+        self.so_line1.discount3 = 50.0
+        self.assertAlmostEqual(self.so_line1.price_subtotal, 75.0)
+        self.assertAlmostEqual(self.order.amount_untaxed, 75.0)
+        self.assertAlmostEqual(self.order.amount_tax, 11.25)
+        # Unset first discount:
+        self.so_line1.discount = 0.0
+        self.assertAlmostEqual(self.so_line1.price_subtotal, 150.0)
+        self.assertAlmostEqual(self.order.amount_untaxed, 150.0)
+        self.assertAlmostEqual(self.order.amount_tax, 22.5)
+        # Set a charge instead:
+        self.so_line1.discount2 = -50.0
+        self.assertAlmostEqual(self.so_line1.price_subtotal, 450.0)
+        self.assertAlmostEqual(self.order.amount_untaxed, 450.0)
+        self.assertAlmostEqual(self.order.amount_tax, 67.5)
+        # set discount_type to additive
+        self.so_line1.discount = 10.0
+        self.so_line1.discount2 = 10.0
+        self.so_line1.discount3 = 10.0
+        self.so_line1.discounting_type = "additive"
+        self.assertAlmostEqual(self.so_line1.price_subtotal, 420.0)
+        self.assertAlmostEqual(self.order.amount_untaxed, 420.0)
+        self.assertAlmostEqual(self.order.amount_tax, 63.0)
+        # set discount over 100%
+        self.so_line1.discount = 30.0
+        self.so_line1.discount2 = 70.0
+        self.so_line1.discount3 = 50.0
+        self.assertAlmostEqual(self.so_line1.price_subtotal, 0.0)
+        self.assertAlmostEqual(self.order.amount_untaxed, 0.0)
+        self.assertAlmostEqual(self.order.amount_tax, 0.0)
+
+    def test_03_sale_order_complex_triple_discount(self):
+        """ Tests on multiple lines """
+        self.so_line1.discount = 50.0
+        self.so_line1.discount2 = 50.0
+        self.so_line1.discount3 = 50.0
+        self.assertAlmostEqual(self.so_line1.price_subtotal, 75.0)
+        self.assertAlmostEqual(self.order.amount_untaxed, 675.0)
+        self.assertAlmostEqual(self.order.amount_tax, 101.25)
+        self.so_line2.discount3 = 50.0
+        self.assertAlmostEqual(self.so_line2.price_subtotal, 300.0)
+        self.assertAlmostEqual(self.order.amount_untaxed, 375.0)
+        self.assertAlmostEqual(self.order.amount_tax, 56.25)
+        self.so_line2.discounting_type = "additive"
+        self.so_line2.discount2 = 10.0
+        self.assertAlmostEqual(self.so_line2.price_subtotal, 240.0)
+        self.assertAlmostEqual(self.order.amount_untaxed, 315.0)
+        self.assertAlmostEqual(self.order.amount_tax, 47.25)
+
+    def test_04_sale_order_triple_discount_invoicing(self):
+        """When a confirmed order is invoiced, the resultant invoice
+        should inherit the discounts"""
+        self.so_line1.discount = 50.0
+        self.so_line1.discount2 = 50.0
+        self.so_line1.discount3 = 50.0
+        self.so_line2.discount3 = 50.0
+        self.order.action_confirm()
```

Comment:
if self.order.state == 'waiting_approval':
        self.order.action_approve()
        self.assertAlmostEqual(self.order.state, 'approved')
        self.order.action_confirm()

Reply:
@ferran-73 This is because of module sale_order_line_discount_validation
It overrides action_confirm method.  Checks for discount limit and adds new states for approval.

Reply:
> ```
> if self.order.state == 'waiting_approval':
>     self.order.action_approve()
>     self.assertAlmostEqual(self.order.state, 'approved')
>     self.order.action_confirm()
> ```

Adding this would mean I should also add the dependency to sale_order_line_discount_validation right? I'm not sure that's ok.

Reply:
Done, big thanks!

Reply:
@ferran-73 it should be like this:
```python
self.order.action_confirm()
if self.order.state == 'waiting_approval': 
    self.order.action_approve() 
    self.assertAlmostEqual(self.order.state, 'approved') 
    self.order.action_confirm()
```